banyand/property/repair_gossip.go

```diff
@@ -0,0 +1,780 @@
+			// there no different nodes, we can skip repair
+			if firstTreeSummaryResp && len(resp.DifferTreeSummary.Nodes) == 0 {
+				return nil
+			}
+			r.handleDifferSummaryFromServer(ctx, stream, resp.DifferTreeSummary, reader, syncShard, clientSlotNodes)
+			firstTreeSummaryResp = false
+		case *propertyv1.RepairResponse_PropertySync:
+			// step 3: keep receiving messages from the server
+			// if the server still sending different nodes, we should keep reading them
+			// if the server sends a PropertySync, we should repair the property and send the newer property back to the server if needed
+			sync := resp.PropertySync
+			updated, newer, err := syncShard.repair(ctx, sync.Id, sync.Property, sync.DeleteTime)
+			if err != nil {
+				r.scheduler.l.Warn().Err(err).Msgf("failed to repair property %s", sync.Id)
+				r.scheduler.metrics.totalRepairFailedCount.Inc(1, request.Group, fmt.Sprintf("%d", request.ShardId))
+			}
+			if updated {
+				r.scheduler.l.Debug().Msgf("successfully repaired property %s", sync.Id)
+				r.scheduler.metrics.totalRepairSuccessCount.Inc(1, request.Group, fmt.Sprintf("%d", request.ShardId))
+				hasPropertyUpdated = true
+				continue
+			}
+			// if the property hasn't been updated, and the newer property is not nil,
+			// which means the property is newer than the server side,
+			if !updated && newer != nil {
+				var p propertyv1.Property
+				err = protojson.Unmarshal(newer.source, &p)
+				if err != nil {
+					r.scheduler.l.Warn().Err(err).Msgf("failed to unmarshal property from db by entity %s", newer.id)
+					continue
+				}
+				// send the newer property to the server
+				err = stream.Send(&propertyv1.RepairRequest{
+					Data: &propertyv1.RepairRequest_PropertySync{
+						PropertySync: &propertyv1.PropertySync{
+							Id:         newer.id,
+							Property:   &p,
+							DeleteTime: newer.deleteTime,
+						},
+					},
+				})
+				if err != nil {
+					r.scheduler.l.Warn().Err(err).Msgf("failed to send newer property sync response to server, entity: %s", newer.id)
+				}
+			}
+		default:
+			r.scheduler.l.Warn().Msgf("unexpected response type: %T, expected DifferTreeSummary or PropertySync", resp)
+		}
+	}
+}
+
+func (r *repairGossipClient) sendPropertyMissing(stream grpclib.BidiStreamingClient[propertyv1.RepairRequest, propertyv1.RepairResponse], entity string) {
+	err := stream.Send(&propertyv1.RepairRequest{
+		Data: &propertyv1.RepairRequest_PropertyMissing{
+			PropertyMissing: &propertyv1.PropertyMissing{
+				Entity: entity,
+			},
+		},
+	})
+	if err != nil {
+		r.scheduler.l.Warn().Err(err).Msgf("failed to send property missing response to client, entity: %s", entity)
+	}
+}
+
+func (r *repairGossipClient) handleDifferSummaryFromServer(
+	ctx context.Context,
+	stream grpclib.BidiStreamingClient[propertyv1.RepairRequest, propertyv1.RepairResponse],
+	differTreeSummary *propertyv1.DifferTreeSummary,
+	reader repairTreeReader,
+	syncShard *shard,
+	clientSlotNodes map[int32]*repairTreeNode,
```

Comment:
You don't need this map as well.

Reply:
We need a sliding window to manage both the summary and the reader, ensuring each is scanned only once.